File: 882/882.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        # convert both strings to dictionaries of character: count
        sDict: dict = self.stringToDict(s)
        tDict: dict = self.stringToDict(t)
        # check that the dictionaries contain the same letters
        return self.dictsAreEqual(sDict, tDict)
        
    def stringToDict(self, s: str) -> dict:
        dic = {}
        for c in s:
            if c not in dic:
                dic[c] = 1
            else:
                dic[c] = dic[c] + 1
    
    def dictsAreEqual(self, s: str, t: str) -> bool:
        for c in s:
            if c not in t:
                return False
            if s[c] != t[c]:
                return False
        return True

File: 882/882.rb
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

# @param {String} s
# @param {String} t
# @return {Boolean}
def is_anagram(s, t)
    if s.length != t.length
        return false
    end
    
    # convert both into a char: count hash
    s_hash = string_to_hash(s)
    t_hash = string_to_hash(t)
        
    # check the counts of the characters
    compare_hashes(s_hash, t_hash)
end

def string_to_hash(s)
    hash = {}
    s.each_char do |c|
        count = 0
        if hash.include?(c)
            count = hash[c]
        end
        hash[c] = count + 1
    end
    hash
end

def compare_hashes(s_hash, t_hash)
    s_hash.each_pair do |s_key, s_value|
        t_value = t_hash[s_key]
        if s_value != t_value
            return false
        end
    end
    true
end
